quinn-proto/src/connection/assembler.rs

```diff
@@ -383,4 +421,35 @@ mod test {
         assert_eq!(x.pop(), Some((0, Bytes::from_static(b"abcdef"))));
         assert_eq!(x.pop(), Some((9, Bytes::from_static(b"jklmno"))));
     }
+
+    #[test]
+    fn unordered_happy_path() {
+        let mut x = Assembler::new();
+        x.insert(0, Bytes::from_static(b"abc"));
+        assert_eq!(x.read_unordered(), Some((0, Bytes::from_static(b"abc"))));
+        assert_eq!(x.read_unordered(), None);
+        x.insert(3, Bytes::from_static(b"def"));
+        assert_eq!(x.read_unordered(), Some((3, Bytes::from_static(b"def"))));
+        assert_eq!(x.read_unordered(), None);
+    }
+
+    #[test]
+    fn unordered_dedup() {
+        let mut x = Assembler::new();
+        x.insert(3, Bytes::from_static(b"def"));
+        assert_eq!(x.read_unordered(), Some((3, Bytes::from_static(b"def"))));
+        assert_eq!(x.read_unordered(), None);
+        x.insert(0, Bytes::from_static(b"abcdefghi"));
```

Comment:
I think large chunks are better than small chunks. If we get a large chunk and we have a small chunk buffered which is a subset of the large chunk, should we keep the larger chunk instead of splitting it up?

Reply:
That would require random-accessing into the `BinaryHeap` somehow to replace the shorter pre-existing chunk, or deferring the deduplication until `read_unordered`-time, which increases memory use. Given that `defragment` will keep a lid on fragmentation overall, and what I estimate to be a quite low incidence of cases where perfect efficiency here would help much, I'm inclined to stick with the easy thing.

Reply:
Thinking on this further, I suspect unordered mode should have a `BTreeMap<u64, Bytes>` rather than the `RangeSet` + `BinaryHeap` arrangement. Then we could optimize for efficient memory layout just as easily as we currently preserve `RangeSet`'s invariants. Removing the first item from a `BTreeMap` to read is awkward, but when we're traversing it multiple times per packet anyway...

I'd rather not block this PR on it, though, since this PR is still a direct improvement on the status quo.

Reply:
Sounds good.